### PR TITLE
Fix constraint validator alias being required

### PR DESCRIPTION
**Pull request by Mx-Glitter**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16805 |
| License | MIT |
| Doc PR | symfony/symfony-docs#5952 |

This is my first contribution, so everything might not be in perfect order.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php [hunk 1]
```diff
@@ -24,6 +24,7 @@ public function process(ContainerBuilder $container)
 
         $validators = array();
         foreach ($container->findTaggedServiceIds('validator.constraint_validator') as $id => $attributes) {
+            $validators[$id] = $id;
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php]: using the id does not make sense. It should use the service class instead (which is what is returned by ``getValidatedBy`` by default

Comment (Triiistan) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/AddConstraintValidatorsPass.php]: I'll modify it but I still have a question for when we deprecate aliases for validators (if we ever do it):

What if there are multiple validator services configured using the same class? This is the same problem as in #16615, so the same choice may apply.

**Comment by Triiistan** [comment]
Do you think Constraint validator aliases should be deprecated here and removed in Symfony 4?

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/AddConstraintValidatorsPassTest.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                'My\Fully\Qualified\Class\Named\Validator1' => 'my_constraint_validator_service1',
+                'my_constraint_validator_alias1' => 'my_constraint_validator_service1',
+                'My\Fully\Qualified\Class\Named\Validator2' => 'my_constraint_validator_service2',
+            ));
+
+        $addConstraintValidatorsPass = new AddConstraintValidatorsPass();
+        $addConstraintValidatorsPass->process($container);
+    }
+
+    public function testThatCompilerPassIsIgnoredIfThereIsNoConstraintValidatoFactoryDefinition()
```
Comment (webmozart) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/AddConstraintValidatorsPassTest.php]: typo here: "ValidatoFactory"

**Comment by webmozart** [comment]
Apart from the typo this looks good to me!

Status: Reviewed

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/AddConstraintValidatorsPassTest.php (hunk 2 above)
Comment (Triiistan) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/Compiler/AddConstraintValidatorsPassTest.php]: Fixed

**Comment by nicolas-grekas** [comment]
If this is a bug fix, it should be merged on 2.3, if it's a new feature it's for master.
@Triiistan I'm not sure the renaming of variables provides any value, nor the $serviceDefinition tmp var. Use `$container->getDefinition($id)->getClass()` instead?

**Comment by Triiistan** [comment]
The underlying issue is [documented](http://symfony.com/doc/current/cookbook/validation/custom_constraint.html#constraint-validators-with-dependencies) as an intended limitation, so I think it can be considered a new feature, and it is backwards compatible (if I'm not mistaken). Isn't 2.8 the right branch for it then, according to the [contribution documentation](http://symfony.com/doc/current/contributing/code/patches.html#choose-the-right-branch)?

The renaming and temporary variable were to make the code consistent with the [FormPass](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FormPass.php#L36-L44) in the same folder. I don't see any problem reverting that if you prefer.

**Comment by xabbuh** [comment]
@Triiistan Thanks for pointing this out. The description you linked to is outdated and must be updated (see symfony/symfony-docs#6045).

**Comment by Triiistan** [comment]
Closed in favor of #17074 on the right branch

**Pull request closed without merging**